src/shims/env.rs

```diff
@@ -150,4 +154,30 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             }
         }
     }
+
+    fn update_environ(&mut self) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
```

Comment:
This function should start by reading the current `environ` and deallocating it.

Reply:
> > For some reason update_environ is not updating environ when setenv or unsetenv are called. Even then it works during initialization.
> 
> That is curious. Could you please add a testcase demonstrating the problem?

sure, check https://github.com/rust-lang/miri/pull/1208/commits/18f03390c8a7423228a3999ffa5d9774076596e9

Reply:
I tried to dealloc it on the last commit but apparently I'm at the wrong pointer. It's causing a null pointer error in runtime. On the other hand I'm not sure how to go to the next nested pointer from there to try and fix it.

Reply:
Well the very first time this gets called, you will indeed see a NULL ptr as that is what `init_extern_statics` puts in. So use `this.is_null(val)` to guard against NULL.